File: 백준/레벨Gold V/1916. 최소비용 구하기/최소비용 구하기.ts
import fs = require('fs');

const lines = fs.readFileSync('/dev/stdin').toString().split('\n');

// 인접리스트 초기화
const N = parseInt(lines[0]);
const M = parseInt(lines[1]);
const graph: [number, number][][] = Array.from({ length: N + 1 }, () => []);

for (let i = 2; i <= M + 1; i++) {
  const [u, v, w] = lines[i].split(' ').map(Number);
  graph[u].push([v, w]);
}

const [S, E] = lines[M + 2].split(' ').map(Number);

// 다익스트라
const costs: number[] = Array(N + 1).fill(Infinity);
costs[S] = 0;

const heap: [number, number][] = [];
heap.push([0, S]);

while (heap.length > 0) {
  heap.sort((a, b) => b[0] - a[0]);
  const [currCost, u] = heap.pop();

  if (currCost > costs[u]) continue;
  if (u === E) break;

  for (const [v, w] of graph[u]) {
    const nextCost = currCost + w;

    if (nextCost < costs[v]) {
      costs[v] = nextCost;
      heap.push([nextCost, v]);
    }
  }
}

console.log(costs[E]);

File: 백준/레벨Gold V/1916. 최소비용 구하기/최소비용 구하기.py
from sys import stdin
from heapq import heappush, heappop

input = stdin.readline

# 인접리스트 초기화
N = int(input().rstrip())
M = int(input().rstrip())
graph = [[] for _ in range(N + 1)]

for _ in range(M):
    u, v, w = map(int, input().split())
    graph[u].append((v, w))

S, E = map(int, input().split())

# 다익스트라
costs = [float('inf')] * (N + 1)
costs[S] = 0

heap = []
heappush(heap, (0, S))

while heap:
    curr_cost, u = heappop(heap)
    
    if curr_cost > costs[u]:
        continue
    if u == E:
        break

    for v, w in graph[u]:
        next_cost = curr_cost + w

        if next_cost < costs[v]:
            costs[v] = next_cost
            heappush(heap, (next_cost, v))

print(costs[E])